cmd/dwirecon.cpp

```diff
@@ -0,0 +1,413 @@
+
+  + Option ("padding", "zero-padding output coefficients to given dimension.")
+    + Argument ("rank").type_integer(0)
+
+  + Option ("complete", "complete (zero-filled) source prediction.")
+
+  + OptionGroup ("CG Optimization options")
+
+  + Option ("tolerance", "the tolerance on the conjugate gradient solver. (default = " + str(DEFAULT_TOL) + ")")
+    + Argument ("t").type_float(0.0, 1.0)
+
+  + Option ("maxiter",
+            "the maximum number of iterations of the conjugate gradient solver. (default = " + str(DEFAULT_MAXITER) + ")")
+    + Argument ("n").type_integer(1)
+
+  + Option ("init",
+            "initial guess of the reconstruction parameters.")
+    + Argument ("img").type_image_in();
+
+}
+
+
+typedef float value_type;
+
+
+
+void run ()
+{
+  // Load input data
+  auto dwi = Image<value_type>::open(argument[0]).with_direct_io({1, 2, 3, 4});
+
+  // Read motion parameters
+  auto opt = get_options("motion");
+  Eigen::MatrixXf motion;
+  if (opt.size())
+    motion = File::Matrix::load_matrix<float>(opt[0][0]);
+  else
+    motion = Eigen::MatrixXf::Zero(dwi.size(3), 6); 
+
+  // Check dimensions
+  if (motion.size() && motion.cols() != 6)
+    throw Exception("No. columns in motion parameters must equal 6.");
+  if (motion.size() && ((dwi.size(3) * dwi.size(2)) % motion.rows()))
+    throw Exception("No. rows in motion parameters does not match image dimensions.");
+
+
+  // Select shells
+  auto grad = DWI::get_DW_scheme (dwi);
+  DWI::Shells shells (grad);
+  shells.select_shells (false, false, false);
+
+  // Read multi-shell basis
+  int lmax = 0;
+  std::vector<Eigen::MatrixXf> rf;
+  opt = get_options("rf");
+  for (size_t k = 0; k < opt.size(); k++) {
+    Eigen::MatrixXf t = File::Matrix::load_matrix<float>(opt[k][0]);
+    if (t.rows() != shells.count())
+      throw Exception("No. shells does not match no. rows in basis function " + opt[k][0] + ".");
+    lmax = std::max(2*(int(t.cols())-1), lmax);
+    rf.push_back(t);
+  }
+
+  // Read slice weights
+  Eigen::MatrixXf W = Eigen::MatrixXf::Ones(dwi.size(2), dwi.size(3));
+  opt = get_options("weights");
+  if (opt.size()) {
+    W = File::Matrix::load_matrix<float>(opt[0][0]);
+    if (W.rows() != dwi.size(2) || W.cols() != dwi.size(3))
+      throw Exception("Weights matrix dimensions don't match image dimensions.");
+  }
+
+  // Read field map and PE scheme
+  opt = get_options("field");
+  bool hasfield = opt.size();
+  auto fieldmap = Image<value_type>();
+  size_t fieldidx = 0;
+  Eigen::MatrixXf PE;
+  if (hasfield) {
+    auto petable = PhaseEncoding::get_scheme(dwi);
+    PE = petable.leftCols<3>().cast<float>();
+    PE.array().colwise() *= petable.col(3).array().cast<float>();
+    // -----------------------  // TODO: Eddy uses a reverse LR axis for storing
+    PE.col(0) *= -1;            // the PE table, akin to the gradient table.
+    // -----------------------  // Fix in the eddy import/export functions in core.
```

Comment:
Need to determine whether this is an artifact of bad metadata handling in *MRtrix3* (#2917), or whether it is in fact the case that the FSL text file formats for indicating the phase encoding table are expected to be based on a negative determinant image. My expectation was that it would be the former. If the comment is true and the issue is the latter, then there will need to be an explicit distinction between storage of the phase encoding table w.r.t. image axes and the import/export of that information from/to FSL.

Reply:
I've removed all of these on the version we're using at KCL now. You can see these changes on my forlk: https://github.com/dchristiaens/shard-recon/compare/master...jdtournier:shard-recon:master

Reply:
@jdtournier Good point. I forgot to merge [your PR](https://github.com/dchristiaens/shard-recon/pull/20) before creating this one. Your last update is that you were going to confirm if the PE direction had been fixed in the main repo. Is this the case? If so, I don't understand why the test I did last week didn't flag a change in PE.

Reply:
Well, all I can say is that the code we're currently running has these lines removed - but I did also have to modify the corresponding `acqp.txt` for it to work correctly... So that would suggest that it isn't fixed after all... 

It terms of testing, I guess what's required is to first run `eddy` to verify the correctness of the `acqp.txt` file, and then run `shardrecon` with same `acqp.txt`, right? Must admit I haven't done that, and given the above, I suspect that will show that the problem persists...

Reply:
1.  Is the derivation of your `acqp.txt` file going through *MRtrix3* processing at all? I'm presuming it's not being loaded from DICOM headers, but are you loading it into an image header and then doing an export to text file? If so then the issues rectified in #2917 could be playing a role.

2. A useful test to do, with this line removed, would be to find the `acqp.txt` that results in `topup` doing the right thing, then convert the NIfTI between LAS and RAS and re-run with the same `acqp.txt`. If it then produces the wrong result, then the FSL interpretation of the phase encoding table is with respect to the image axes, as has always been my understanding. If it produces the correct result, then FSL expects the phase encoding table to be defined with respect to its internal left-handed coordinate system, and our command-line interface for phase encoding data handling has to be changed.